### PR TITLE
[Scala 3] Add support for given definitions

### PR DESCRIPTION
**Pull request by tgodzik**
Given is a replacement for implicits that was added for Scala 3. It can exist in a couple of forms:
- given declaration `given intOrd: Ord[Int]`
- given alias `given intOrd: Ord[Int] = existingOrd`
- given definition `given intOrd: Ord[Int] with { def compare = ??? }`

Alias and definition can have the name skipped. Givens can additionally be typed and have using parameters.

More info at: https://dotty.epfl.ch/docs/reference/contextual/givens.html

~~Created a draft PR because we are blocked on https://github.com/scalameta/scalameta/pull/2220, but since this is a large PR opened it early if anyone is willing to take a look.~~ It's released in `4.4.8`

### CODE REVIEW

File: scalafmt-tests/src/test/resources/scala3/OpaqueType.stat [hunk 1]
```diff
@@ -15,8 +15,8 @@ maxColumn = 15
 ===
 opaque type VeryLongClassName <: A & B = AB
 >>>
-opaque type VeryLongClassName <: A & B =
-  AB
+opaque type VeryLongClassName <: A &
```
Comment (tgodzik) [inline scalafmt-tests/src/test/resources/scala3/OpaqueType.stat:18]: So this change is caused by updating the scalameta version.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 2]
```diff
@@ -414,6 +414,14 @@ class FormatOps(
     template.tokens.find(x => x.is[T.LeftBrace] && owners(x) == template)
   }
 
+  def templateWith(template: Template): Option[Token] = {
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:417]: 1. i think that method `templateCurly` on line 409 is unused, i'd remove it
2. this method is used in only one place; should we inline it there?

File: scalafmt-tests/src/test/resources/scala3/OpaqueType.stat (hunk 1 above)
Comment (kitbellew) [reply scalafmt-tests/src/test/resources/scala3/OpaqueType.stat:18]: 1. should scalameta update be a separate pull request?
2. should we fix something about this formatting?

**Review by kitbellew — CHANGES_REQUESTED** [review]

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 3]
```diff
@@ -414,6 +414,14 @@ class FormatOps(
...
+    val tokens = template.tokens
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:418]: That's what I started with 😅 but it's not available in Scala 2.11. That brings me to another question... should we actually drop 2.11 support? Not sure if it's still needed. We can run with scala211 dialect anyway.

[the remark this answers is not in the document]
Comment (kitbellew) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:418]: i am not talking about `Seq.findLast` which is available in 2.13. i implemented our own method, also called `findLast`, within `TokenOps`.

[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala:418]: Ach! Sorry, I started looking at the mobile app and did not notice that you mentioned TokenOps.

File: scalafmt-tests/src/test/resources/scala3/OpaqueType.stat (hunk 1 above)
Comment (tgodzik) [reply scalafmt-tests/src/test/resources/scala3/OpaqueType.stat:18]: We need the scalameta update for this PR, but might make sense to update it in a separate PR. I will create one and address the test.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/RedundantParens.scala [hunk 4]
```diff
@@ -22,6 +22,8 @@ class RedundantParens(implicit ctx: RewriteCtx) extends RewriteSession {
 
       case t: Defn.Def => maybeRemove(t.body)
 
+      case t: Defn.GivenAlias => maybeRemove(t.body)
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/rewrite/RedundantParens.scala:25]: We do here: 

https://github.com/scalameta/scalafmt/pull/2332/commits/8a195be55648a7705a6d49d91ae761567f56a444#diff-8b41c5f61c6bc86714f846225ae0f4325d8f22aafccba6d6746a6ac7b831818dR202

Given alias is just a given with an equals.

File: scalafmt-tests/src/test/resources/scala3/OpaqueType.stat (hunk 1 above)
Comment (tgodzik) [reply scalafmt-tests/src/test/resources/scala3/OpaqueType.stat:18]: Opened a PR: https://github.com/scalameta/scalafmt/pull/2351

We can resolve that one first.

Comment (tgodzik) [reply scalafmt-tests/src/test/resources/scala3/OpaqueType.stat]: The other PR is merged :tada:

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: Fixed!

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala [hunk 5]
```diff
@@ -948,7 +956,12 @@ class FormatOps(
       indent: Int
   ): Seq[Split] = {
     val hasSelfAnnotation = template.self.tokens.nonEmpty
-    val expire = templateCurly(template).getOrElse(template.tokens.last)
+    val expire =
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: Good idea, changed! A lot  less code now :sweat_smile:

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala [hunk 6]
```diff
@@ -1455,12 +1458,18 @@ class Router(formatOps: FormatOps) {
         rightOwner match {
           // something like new A with B with C
           case template: Template if template.parent.exists { p =>
-                p.is[Term.New] || p.is[Term.NewAnonymous]
+                p.is[Term.New] || p.is[Term.NewAnonymous] || p.is[Defn.Given]
               } =>
+            // when given is used it should have increased indentation to work with optional braces
+            val lastIdentedToken =
```
[the remark this answers is not in the document]
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/Router.scala]: 1. Yes! Good point.
2. Changed I hope everywhere.

File: scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala (hunk 2 above)
Comment (tgodzik) [reply scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatOps.scala]: 1. Removed.
2. Inlined.
:tada:

File: scalafmt-tests/src/test/resources/scala3/Given.stat [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+}
+>>>
+given a: A
+  with Alpha
+  with Beta
+  with Gamma
+  with Delta
+  with Epsilon
+  with {
+    def a = ???
```
Comment (kitbellew) [inline scalafmt-tests/src/test/resources/scala3/Given.stat:137]: just a question: is this better (or more consistent with scala3's significant indentation) than if we had reduced indentation inside `{}`?

Comment (tgodzik) [reply scalafmt-tests/src/test/resources/scala3/Given.stat:137]: Actually, let me check. Maybe it's enough to just have the same level of indentation as 'with'.

Comment (tgodzik) [reply scalafmt-tests/src/test/resources/scala3/Given.stat]: Removed the increased indentation. It's readable enough without it. What to you think?

**Review by kitbellew — APPROVED** [review]
looks like the code has been simplified, too, removing the alternative indentation token logic.

File: scalafmt-tests/src/test/resources/scala3/Given.stat (hunk 7 above)
Comment (kitbellew) [reply scalafmt-tests/src/test/resources/scala3/Given.stat]: looks good to me

**Pull request merged**